File: models/port_reservation.rb
class PortReservation < ActiveRecord::Base
  belongs_to :node
  belongs_to :transaction_chain

  def self.reserve(node, addr, chain)
    self.where(node: node, addr: nil).order('port ASC').limit(10).each do |r|
      begin
        r.update!(addr: addr, transaction_chain: chain)
        return r

      rescue ActiveRecord::RecordNotUnique
        next
      end
    end

    fail 'no port available'
  end

  def free
    update!(addr: nil, transaction_chain: nil)
  end
end

File: core/apps/vpsadmin_persistence/lib/vpsadmin_persistence/schema/port_reservation.ex
defmodule VpsAdmin.Persistence.Schema.PortReservation do
  use VpsAdmin.Persistence.Schema

  schema "port_reservations" do
    field(:addr, :string)
    field(:port, :integer)
    belongs_to(:node, Schema.Node)
    belongs_to(:transaction_chain, Schema.TransactionChain)
  end
end
